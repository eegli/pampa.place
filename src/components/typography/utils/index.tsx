import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
