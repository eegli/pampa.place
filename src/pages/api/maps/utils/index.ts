export {apiData as data} from '@/config/maps';
import {NextApiHandler} from 'next';

type ErrorResponse = {
  error: string;
};

type DataResponse<T> = T extends unknown[]
  ? {
      data: T;
    }
  : {
      data: T[];
    };

export type ApiJSONHandler<T = unknown> = NextApiHandler<
  DataResponse<T> | ErrorResponse
>;
export type ApiTextHandler = NextApiHandler<string> | ErrorResponse;

export function nthQuery<T>(arg: T | T[], index = 0): T | undefined {
  return Array.isArray(arg) ? arg[index] : arg;
}
