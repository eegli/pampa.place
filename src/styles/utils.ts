import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };
