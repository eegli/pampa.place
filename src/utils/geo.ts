import {LatLngLiteral} from '@/config/types';
import Tdistance from '@turf/distance';
import {BBox, MultiPolygon, point, Polygon, Units} from '@turf/helpers';
import TpointsWithinPolygon from '@turf/points-within-polygon';
import {randomPoint as TrandomPoint} from '@turf/random';

export function randomPointInMap(
  bbox: BBox,
  polygon: Polygon | MultiPolygon
): LatLngLiteral {
  do {
    const random = TrandomPoint(20, {bbox});
    const ptsWithin = TpointsWithinPolygon(random, polygon);
    if (ptsWithin.features.length) {
      const pt = ptsWithin.features[0].geometry.coordinates;
      return {lng: pt[0], lat: pt[1]};
    }
  } while (true);
}

// Area in square kilometers, distance in meters
// https://www.desmos.com/calculator/xlzbhq4xm0
export function calcScore(area: number, dist: number): number {
  if (dist < 0) return 0;
  const c = Math.log(area * dist + 1) * Math.sqrt(area * dist);
  const score = 5000 * Math.E ** -(dist / c);
  return Math.round(score);
}
// Returns distance between two points in kilometers

export function calcDist(
  p1: LatLngLiteral,
  p2: LatLngLiteral,
  units: Units = 'meters'
): number {
  const from = point([p1.lng, p1.lat]);
  const to = point([p2.lng, p2.lat]);
  return Tdistance(from, to, {units});
}
