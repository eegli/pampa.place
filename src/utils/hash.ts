export async function sha256(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message);
  const hashBuffer = await crypto.subtle.digest('SHA-256', msgUint8);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
  return hashHex;
}

export function id(): string {
  const array = new Uint32Array(8);
  window.crypto.getRandomValues(array);
  let str = '';
  for (let i = 0; i < array.length; i++) {
    str += (i < 2 || i > 5 ? '' : '-') + array[i].toString(16).slice(-4);
  }
  return str;
}
