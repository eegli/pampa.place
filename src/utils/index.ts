// Area in square kilometers, distance in meters
import { LatLngLiteral } from '@config';
import * as turf from '@turf/turf';

// https://www.desmos.com/calculator/xlzbhq4xm0
export function calculateScore(area: number, dist: number) {
  if (dist < 0) return 0;
  const c =
    Math.log(area + 1) * Math.sqrt(dist * area) + Math.log(dist ** 2 + 1);
  const score = 5000 * Math.E ** -(dist / c);

  return Math.round(score);
}
// Returns distance between two points in kilometers

export function calculateDistance(
  p1: LatLngLiteral,
  p2: LatLngLiteral,
  units: turf.Units = 'kilometers'
) {
  const from = turf.point([p1.lng, p1.lat]);
  const to = turf.point([p2.lng, p2.lat]);
  return turf.distance(from, to, { units });
}
// Formats distance
export function formatDist(km: number, toFixed = 1) {
  if (km < 0) return '-';
  if (km > 1000) {
    return (km / 1000).toFixed(toFixed) + ' m';
  }
  return km.toFixed(toFixed) + ' km';
}

export function max(num: number, limit: number) {
  return num >= limit ? limit : num;
}

export function id(): string {
  const array = new Uint32Array(8);
  window.crypto.getRandomValues(array);
  let str = '';
  for (let i = 0; i < array.length; i++) {
    str += (i < 2 || i > 5 ? '' : '-') + array[i].toString(16).slice(-4);
  }
  return str;
}
