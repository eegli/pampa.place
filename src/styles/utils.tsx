import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

type KeyedColorFadeProps = [number, string][];

export const keyedColorFade =
  (prop: string) =>
  (...steps: KeyedColorFadeProps) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
