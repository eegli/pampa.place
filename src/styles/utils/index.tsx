import createCache from '@emotion/cache';
import Box from '@mui/material/Box';
import {keyframes} from '@mui/material/styles';

const isBrowser = typeof document !== 'undefined';

// On the client side, Create a meta tag at the top of the <head> and set it as insertionPoint.
// This assures that MUI styles are loaded first.
// It allows developers to easily override MUI styles with other styling solutions, like CSS modules.
export function createEmotionCache() {
  let insertionPoint;

  if (isBrowser) {
    const emotionInsertionPoint = document.querySelector<HTMLMetaElement>(
      'meta[name="emotion-insertion-point"]'
    );
    insertionPoint = emotionInsertionPoint ?? undefined;
  }

  return createCache({key: 'mui-style', insertionPoint});
}

export const keyedColorFade =
  (prop: string) =>
  (...steps: [number, string][]) => {
    const gradient = steps
      .map(([perc, color]) => `${perc}% {${prop}: ${color}}`)
      .join(' ');
    return keyframes(gradient);
  };

export const colorize = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
