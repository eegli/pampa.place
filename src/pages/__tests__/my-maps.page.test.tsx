import {createMockState, createMockStore, render, screen} from '@/tests/utils';
import {MyMapsPage} from '../poc/my-maps.page';

beforeEach(() => {
  jest.clearAllMocks();
});

function getNameInput() {
  return screen.queryByLabelText(/map name/i);
}

describe('My maps page', () => {
  it(`renders`, () => {
    const state = createMockState();
    const store = createMockStore(state);
    render(<MyMapsPage />, store);
    expect(getNameInput()).toMatchSnapshot();
  });
});
