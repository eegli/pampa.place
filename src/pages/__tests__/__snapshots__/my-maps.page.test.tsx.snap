// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My maps page renders 1`] = `
<input
  aria-invalid="false"
  class="MuiOutlinedInput-input MuiInputBase-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
  id="mui-1"
  placeholder="Add a display name for the map"
  type="text"
  value=""
/>
`;
